x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -364,13 +425,16 @@ private void loadApiKeyAndValidateCredentials(ThreadContext ctx, ApiKeyCredentia
                         response.getSourceAsBytesRef(), XContentType.JSON)) {
                         apiKeyDoc = ApiKeyDoc.fromXContent(parser);
                     }
-                    validateApiKeyCredentials(docId, apiKeyDoc, credentials, clock, ActionListener.delegateResponse(listener, (l, e) -> {
-                        if (ExceptionsHelper.unwrapCause(e) instanceof EsRejectedExecutionException) {
-                            listener.onResponse(AuthenticationResult.terminate("server is too busy to respond", e));
-                        } else {
-                            listener.onFailure(e);
+                    if (checkpoint != null) {
+                        final CachedApiKeyDoc cachedApiKeyDoc = apiKeyDoc.toCachedApiKeyDoc();
+                        if (checkpoint.put(docId, cachedApiKeyDoc)) {
```

Comment:
It's a bit hard to understand the reasoning for the "consistent cache" here.

My understanding is that you're trying to guard against a doc update (a call to invalidate a key), followed by a cache clear call (they are currently separate endpoints), **after** the doc has been retrieved by the get, but before the handler is called. IMO, this is a narrow time window for this big of a complication (the `ConsistentCache` complication).

Alternatively, what do you say you make the **current** invalidation counter part of the cached **value** (the `CachedApiKeyDoc` instance). In this case, you always populate the cache values with the current invalidation counter, but you never use a cache entry with an older invalidation counter compared to the current (you remove the entry and try the get).

Reply:
The `ConsistentCache` is an attempt to extract a pattern from existing code so that caching behaviour can be reused in different places. The [`CompositeRolesStore`](https://github.com/elastic/elasticsearch/blob/7.9/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L301-L312) is an example of this pattern, which is also applied for [`NativePrivilegeStore`](https://github.com/elastic/elasticsearch/blob/7.9/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java#L432-L439) in my previous PR (#55836) for caching application privileges. The complexity is real but it is not new. It is exactly because of the complexity, I am attempting to extract the logic in a separate class so it can be reused, e.g. replacing the existing caching code of `NativePrivilegeStore` in future PRs.

The purpose of this piece of code is to minimize the possibility of caching stale results and invalidate it as soon as possible if we do end up caching it. Tim and I had some [discussion](https://github.com/elastic/elasticsearch/pull/55836#discussion_r416298343) about how it works. Unfortunately, I just realised that I didn't add a comment after our discussion so that the reasonsing is not obvious. I will try to capture it here. In summary, both the `AtomicLong` and `ReadWriteLock` are necessary due to the aysnc and non-atomic nature of operations in the system. The problem arises because of following few things:
* The read and write calls to the security index are independant. 
* The clear cache call is independant as well. This include both standalone clear cache call initiated by the users or the ones as consequence of write calls.
* `AtomicLong#get` and any actions guarded by it are not an atomic group.

The `ReadWriteLock` helps make above async stuff easier to reason about and minimize the possibility of stable results being cache. We can use following scenario as an example:

A read successfully retrieves an document from the security index. But before the code is done with retrieved document, an write call happens. The write call basically makes the last read document stale. So we do not want to cache the retrieved document. The write call fires a cache invalidation request, which is currently inflight. Before actuall caching the document, the read thread checks whether invalidation has happened with following code:
```java
if (invalidationCounter == numInvalidation.get()) {
    cacheTheDocument();
}
```
However, even if the check passes, it is still possible that the invalidation thread gets executed immedilatey after the check and before the actual caching call. So the read thread will end up caching a stale document. The chance for this happen may be low, but it is hard to reason about because it is a racing condition. When it happened, it is likely an API key is invalidated, but because it is errorneously cached, subsequent requests will only see the cached result.

When we add the `ReadWriteLock`, the code becomes:
```java
try (ReleasableLock ignored = invalidationReadLock.acquire()) {
    if (invalidationCounter == numInvalidation.get()) {
        cacheTheDocument();
    }
}
```
It helps with above issue because: If the read thread reaches the critical region first and acquires the lock, it can be sure that the counter will not suddenly change again when `cacheTheDocument` is executed. In another word, invalidation thread will be put on hold because of the locking. So even if the read thread does cache the stale document, once it releases the lock, invalidation thread will acquire it and invalidate the document momentarily. Thus minimises the possibility of caching stale document.

Witout the `ReadWriteLock`, it does not matter where you place the `AtomicLong` counter because the real problem is because anything guarded by the counter cannot be guarantteed to finish execution before the counter changes again. And this could happen regardless of where the code is.